web/pandas/pdeps/0013-standardize-apply.md

```diff
@@ -0,0 +1,333 @@
+
+```python
+>>> small_ser.apply(lambda x: x.sum())
+6
+>>> small_ser.apply([lambda x: x.sum()])
+<lambda>    6
+dtype: int64
+```
+
+### 3. Functions in `Series.apply`
+
+The `Series.apply` doc string have examples with using lambdas, but using lambdas in `Series.apply` is often a bad practices because of bad performance:
+
+```python
+>>> %timeit large_ser.apply(lambda x: x + 1)
+24.1 ms ± 88.8 µs per loop
+```
+
+Currently, `Series` does not have a method that makes a callable operate on a series' data. Instead users need to use `Series.pipe` for that operation in order for the operation to be efficient:
+
+```python
+>>> %timeit large_ser.pipe(lambda x: x + 1)
+44 µs ± 363 ns per loop
+```
+
+(The reason for the above performance differences is that apply gets called on each single element, while `pipe` calls `x.__add__(1)`, which operates on the whole array).
+
+Note also that `.pipe` operates on the `Series` while `apply`currently operates on each element in the data, so there is some differences that may have some consequence in some cases.
+
+This PDEP proposes that callables will be applies to the whole `Series`, so we in the future `Series.apply` will be as fast as `Series.pipe`.
+
+### 4. ufuncs in `Series.apply` vs. normal functions
+
+Performance-wise, ufuncs are fine in `Series.apply`, but non-ufunc functions are not:
+
+```python
+>>> %timeit large_ser.apply(np.sqrt)
+71.6 µs ± 1.17 µs per loop
+>>> %timeit large_ser.apply(lambda x:np.sqrt(x))
+63.6 ms ± 1.03 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
+```
+
+It is difficult to understand why ufuncs are fast in `apply`, while other callables are slow in `apply` (answer: it's because ufuncs operate on the whole Series, while other callables operate elementwise).
+
+This PDEP proposes that callables will be applies to the whole `Series`, so we in the future non-ufunc functions in `Series.apply` will be as fast as ufuncs.
+
+### 5. callables in `Series.apply` is slow, callables in `DataFrame.apply` is fast
+
+Above it was shown that using (non-ufunc) callables in `Series.apply` is bad performance-wise. OTOH using them in `DataFrame.apply` is fine:
+
+```python
+>>> %timeit large_ser.apply(lambda x: x + 1)
+24.3 ms ± 24 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
+>>> %timeit pd.DataFrame(large_ser).apply(lambda x: x + 1)
+160 µs +- 1.17 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
+```
+
+Having callables being fast to use in the `DataFrame.apply` method, but slow in `Series.apply` is confusing for users.
+
+This PDEP proposes that callables will be applies to the whole `Series`, so we in the future `Series.apply` will be as fast as `DataFrame.apply` already is.
+
+### 6. callables in `Series.apply` may fail, while callables in `DataFrame.apply` do not and vica versa
+
+```python
+>>> ser.apply(lambda x: x.sum())
+AttributeError: 'int' object has no attribute 'sum'
+>>> pd.DataFrame(ser).apply(lambda x: x.sum())
+0    6
+dtype: int64
+```
```

Comment:
You should show the reverse example:
```python
In [27]: small_ser.apply(lambda x: x+3 if x<2 else x*10)
Out[27]:
0     4
1    20
2    30
dtype: int64

In [28]: pd.DataFrame(small_ser).apply(lambda x: x+3 if x<2 else x*10)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-28-7950464b3dd1> in <module>
----> 1 pd.DataFrame(small_ser).apply(lambda x: x+3 if x<2 else x*10)

C:\Anaconda3\lib\site-packages\pandas\core\frame.py in apply(self, func, axis, raw, result_type, args, **kwargs)
   8734             kwargs=kwargs,
   8735         )
-> 8736         return op.apply()
   8737
   8738     def applymap(

C:\Anaconda3\lib\site-packages\pandas\core\apply.py in apply(self)
    686             return self.apply_raw()
    687
--> 688         return self.apply_standard()
    689
    690     def agg(self):

C:\Anaconda3\lib\site-packages\pandas\core\apply.py in apply_standard(self)
    810
    811     def apply_standard(self):
--> 812         results, res_index = self.apply_series_generator()
    813
    814         # wrap results

C:\Anaconda3\lib\site-packages\pandas\core\apply.py in apply_series_generator(self)
    826             for i, v in enumerate(series_gen):
    827                 # ignore SettingWithCopy here in case the user mutates
--> 828                 results[i] = self.f(v)
    829                 if isinstance(results[i], ABCSeries):
    830                     # If we have a view on v, we need to make a copy because

<ipython-input-28-7950464b3dd1> in <lambda>(x)
----> 1 pd.DataFrame(small_ser).apply(lambda x: x+3 if x<2 else x*10)

C:\Anaconda3\lib\site-packages\pandas\core\generic.py in __nonzero__(self)
   1536     def __nonzero__(self):
   1537         raise ValueError(
-> 1538             f"The truth value of a {type(self).__name__} is ambiguous. "
   1539             "Use a.empty, a.bool(), a.item(), a.any() or a.all()."
   1540         )

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```